File: crud-framework-web/test/java/com/antelopesystem/crudframework/fieldmapper/transformer/TransformerUnitTests.java
package com.antelopesystem.crudframework.fieldmapper.transformer;

import com.antelopesystem.crudframework.fieldmapper.transformer.annotation.EnumType;
import org.junit.Test;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import static org.junit.Assert.assertEquals;

public class TransformerUnitTests {

	@Test
	public void testCommaDelimitedStringToListTransformer() {
		String testString = "var1,var2,var3,var4";
		List<String> expectedOutcome = Arrays.asList("var1", "var2", "var3", "var4");
		CommaDelimitedStringToListTransformer transformer = new CommaDelimitedStringToListTransformer();
		List<String> outcome = transformer.transform(TestPojo.getField("testString"), TestPojo.getField("testStringList"), testString);
		assertEquals(expectedOutcome, outcome);
	}

	@Test
	public void testStringListToCommaDelimitedStringTransformer() {
		List<String> testStringList = Arrays.asList("var1", "var2", "var3", "var4");
		String expectedOutcome = "var1,var2,var3,var4";
		StringListToCommaDelimitedStringTransformer transformer = new StringListToCommaDelimitedStringTransformer();
		String outcome = transformer.transform(TestPojo.getField("testStringList"), TestPojo.getField("testString"), testStringList);
		assertEquals(expectedOutcome, outcome);
	}

	@Test
	public void testCurrencyDoubleToLongTransformer() {
		double testDouble = 112.2;
		long expectedOutcome = 11220;
		CurrencyDoubleToLongTransformer transformer = new CurrencyDoubleToLongTransformer();
		long outcome = transformer.transform(TestPojo.getField("testDouble"), TestPojo.getField("testLong"), testDouble);
		assertEquals(expectedOutcome, outcome);
	}

	@Test
	public void testLongToCurrencyDoubleTransformer() {
		long testLong = 11220;
		double expectedOutcome = 112.2;
		LongToCurrencyDoubleTransformer transformer = new LongToCurrencyDoubleTransformer();
		double outcome = transformer.transform(TestPojo.getField("testLong"), TestPojo.getField("testDouble"), testLong);
		assertEquals(expectedOutcome, outcome, 0.0);
	}

	@Test
	public void testDateToLongTransformer() {
		Date testDate = new Date(100000);
		long expectedOutcome = testDate.getTime();
		DateToLongTransformer transformer = new DateToLongTransformer();
		long outcome = transformer.transform(TestPojo.getField("testDate"), TestPojo.getField("testLong"), testDate);
		assertEquals(expectedOutcome, outcome);
	}

	@Test
	public void testLongToDateTransformer() {
		long testLong = 100000;
		Date expectedOutcome = new Date(testLong);
		LongToDateTransformer transformer = new LongToDateTransformer();
		Date outcome = transformer.transform(TestPojo.getField("testLong"), TestPojo.getField("testDate"), testLong);
		assertEquals(expectedOutcome, outcome);
	}

	@Test
	public void testToStringTransformer() {
		Integer testInt = 130405;
		String expectedOutcome = testInt.toString();
		ToStringTransformer transformer = new ToStringTransformer();
		String outcome = transformer.transform(TestPojo.getField("testInt"), TestPojo.getField("testString"), testInt);
		assertEquals(expectedOutcome, outcome);
	}

	@Test
	public void testDefaultTransformer() {
		assertEquals(1L, new DefaultTransformer().transform(TestPojo.getField("testLong"), TestPojo.getField("testLong"), 1L));
	}

//	@Test
//	public void testBaseEntityToIdTransformer() {
//		BaseEntity testBaseEntity = new BaseEntity() {
//			@Override
//			public Long getId() {
//				return 55L;
//			}
//		};
//
//		assertEquals(55L, (long) new BaseEntityToIdTransformer().transform(TestPojo.getField("testBaseEntity"), TestPojo.getField("testLong"), testBaseEntity));
//	}

	@Test
	public void testCommaDelimitedStringToEnumListTransformer() {
		String testString = "First,Third";
		List<TestEnum> expectedOutcome = Arrays.asList(TestEnum.First, TestEnum.Third);

		assertEquals(expectedOutcome, new CommaDelimitedStringToEnumListTransformer().transform(TestPojo.getField("testString"), TestPojo.getField("testEnumList"), testString));
	}

	@Test
	public void testEnumListToCommaDelimitedString() {
		List<TestEnum> testEnumList = Arrays.asList(TestEnum.Second, TestEnum.First, TestEnum.Second);
		String expectedOutcome = "Second,First,Second";
		assertEquals(expectedOutcome, new EnumListToCommaDelimitedString().transform(TestPojo.getField("testEnumList"), TestPojo.getField("testString"), testEnumList));
	}

	@Test
	public void testEnumToStringTransformer() {
		TestEnum testEnum = TestEnum.Third;
		String expectedOutcome = "Third";
		EnumToStringTransformer enumToStringTransformer = new EnumToStringTransformer();
		assertEquals(expectedOutcome, enumToStringTransformer.transform(TestPojo.getField("testEnum"), TestPojo.getField("testString"), testEnum));
	}

	static class TestPojo {

		private String testString;

		private List<String> testStringList;

		private Long testLong;

		private Double testDouble;

		private Integer testInt;

		private Date testDate;

		private TestEnum testEnum = TestEnum.Third;

		@EnumType(TestEnum.class)
		private List<TestEnum> testEnumList;

		public static Field getField(String name) {
			try {
				return TestPojo.class.getDeclaredField(name);
			} catch(Exception e) {
				throw new RuntimeException(e);
			}
		}
	}

	private enum TestEnum {
		First, Second, Third
	}


}

File: crud-framework-fieldmapper/src/test/java/com/antelopesystem/crudframework/fieldmapper/transformer/TransformerUnitTests.kt
package com.antelopesystem.crudframework.fieldmapper.transformer

import com.antelopesystem.crudframework.fieldmapper.annotation.StringTrimLength
import com.antelopesystem.crudframework.fieldmapper.transformer.annotation.EnumType
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.api.expectThrows
import strikt.assertions.isEqualTo
import strikt.assertions.isNull
import java.lang.reflect.Field
import java.util.*

class TransformerUnitTests {
    @Test
    fun testCommaDelimitedStringToListTransformer() {
        val testString = "var1,var2,var3,var4"
        val expectedOutcome = Arrays.asList("var1", "var2", "var3", "var4")
        val transformer = CommaDelimitedStringToListTransformer()
        val outcome: List<String?> = transformer.transform(TestPojo.getField("testString"), TestPojo.getField("testStringList"), testString, TestPojo.INSTANCE, TestPojo.INSTANCE) as List<String?>
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testStringListToCommaDelimitedStringTransformer() {
        val testStringList = Arrays.asList("var1", "var2", "var3", "var4")
        val expectedOutcome = "var1,var2,var3,var4"
        val transformer = StringListToCommaDelimitedStringTransformer()
        val outcome = transformer.transform(TestPojo.getField("testStringList"), TestPojo.getField("testString"), testStringList, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testCurrencyDoubleToLongTransformer() {
        val testDouble = 112.2
        val expectedOutcome: Long = 11220
        val transformer = CurrencyDoubleToLongTransformer()
        val outcome = transformer.transform(TestPojo.getField("testDouble"), TestPojo.getField("testLong"), testDouble, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testLongToCurrencyDoubleTransformer() {
        val testLong: Long = 11220
        val expectedOutcome = 112.2
        val transformer = LongToCurrencyDoubleTransformer()
        val outcome = transformer.transform(TestPojo.getField("testLong"), TestPojo.getField("testDouble"), testLong, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)

    }

    @Test
    fun testDateToLongTransformer() {
        val testDate = Date(100000)
        val expectedOutcome = testDate.time
        val transformer = DateToLongTransformer()
        val outcome = transformer.transform(TestPojo.getField("testDate"), TestPojo.getField("testLong"), testDate, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testLongToDateTransformer() {
        val testLong: Long = 100000
        val expectedOutcome = Date(testLong)
        val transformer = LongToDateTransformer()
        val outcome = transformer.transform(TestPojo.getField("testLong"), TestPojo.getField("testDate"), testLong, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testToStringTransformer() {
        val testInt = 130405
        val expectedOutcome = testInt.toString()
        val transformer = ToStringTransformer()
        val outcome = transformer.transform(TestPojo.getField("testInt"), TestPojo.getField("testString"), testInt, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testDefaultTransformer() {
        val outcome = DefaultTransformer().transform(TestPojo.getField("testLong"), TestPojo.getField("testLong"), 1L, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(1L)
    }

    @Test
    fun testCommaDelimitedStringToEnumListTransformer() {
        val testString = "First,Third"
        val expectedOutcome = Arrays.asList(TestEnum.First, TestEnum.Third)
        val outcome = CommaDelimitedStringToEnumListTransformer().transform(TestPojo.getField("testString"), TestPojo.getField("testEnumList"), testString, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testEnumListToCommaDelimitedString() {
        val testEnumList = Arrays.asList(TestEnum.Second, TestEnum.First, TestEnum.Second)
        val expectedOutcome = "Second,First,Second"
        val outcome = EnumListToCommaDelimitedString().transform(TestPojo.getField("testEnumList"), TestPojo.getField("testString"), testEnumList, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun testEnumToStringTransformer() {
        val testEnum = TestEnum.Third
        val expectedOutcome = "Third"
        val enumToStringTransformer = EnumToStringTransformer()
        val outcome = enumToStringTransformer.transform(TestPojo.getField("testEnum"), TestPojo.getField("testString"), testEnum, TestPojo.INSTANCE, TestPojo.INSTANCE)

        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    @Test
    fun `test StringTrimTransformer fails on missing @StringTrimLength annotation`() {
        val transformer = StringTrimTransformer()
        expectThrows<IllegalStateException> {
            transformer.transform(TestPojo.getField("testString"), TestPojo.getField("testString"), "test", TestPojo.INSTANCE, TestPojo.INSTANCE)
        }
    }

    @Test
    fun `test StringTrimTransformer returns null on null`() {
        val transformer = StringTrimTransformer()
        val outcome = transformer.transform(TestPojo.getField("testStringTrim"), TestPojo.getField("testStringTrim"), null, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isNull()
    }

    @Test
    fun `test StringTrimTransformer happy flow`() {
        val transformer = StringTrimTransformer()
        val input = "testing123"
        val expectedOutcome = "testi"
        val outcome = transformer.transform(TestPojo.getField("testStringTrim"), TestPojo.getField("testStringTrim"), input, TestPojo.INSTANCE, TestPojo.INSTANCE)
        expectThat(outcome)
            .isEqualTo(expectedOutcome)
    }

    internal class TestPojo {
        private val testString: String? = null
        private val testStringList: List<String>? = null
        private val testLong: Long? = null
        private val testDouble: Double? = null
        private val testInt: Int? = null
        private val testDate: Date? = null
        private val testEnum = TestEnum.Third

        @EnumType(TestEnum::class)
        private val testEnumList: List<TestEnum>? = null

        @StringTrimLength(5)
        private val testStringTrim: String? = null

        companion object {
            val INSTANCE = TestPojo()
            fun getField(name: String?): Field {
                return try {
                    TestPojo::class.java.getDeclaredField(name)
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            }
        }
    }

    private enum class TestEnum {
        First, Second, Third
    }
}